sdk/python/feast/pyspark/batch_retrieval_job.py

```diff
@@ -0,0 +1,203 @@
+
+    Returns:
+        DataFrame: Join result.
+
+    """
+    entity_with_id = entity.withColumn("_row_nr", monotonically_increasing_id())
+
+    feature_event_timestamp = f"{feature_prefix}event_timestamp"
+    feature_created_timestamp = f"{feature_prefix}created_timestamp"
+
+    projection = [
+        col(col_name).alias(f"{feature_prefix}{col_name}")
+        for col_name in join_keys + features + ["event_timestamp", "created_timestamp"]
+    ]
+
+    selected_feature_table = feature_table.select(projection)
+
+    join_cond = (
+        entity_with_id.event_timestamp
+        >= selected_feature_table[feature_event_timestamp]
+    )
+    if max_age:
+        join_cond = join_cond & (
+            selected_feature_table[feature_event_timestamp]
+            >= entity_with_id.event_timestamp - expr(f"INTERVAL {max_age}")
+        )
+
+    for key in join_keys:
+        join_cond = join_cond & (
+            entity_with_id[key] == selected_feature_table[f"{feature_prefix}{key}"]
+        )
+
+    conditional_join = entity_with_id.join(
+        selected_feature_table, join_cond, "leftOuter"
+    )
+    for key in join_keys:
+        conditional_join = conditional_join.drop(
+            selected_feature_table[f"{feature_prefix}{key}"]
+        )
+
+    window = Window.partitionBy("_row_nr", *join_keys).orderBy(
+        col(feature_event_timestamp).desc(), col(feature_created_timestamp).desc()
+    )
+    filter_most_recent_feature_timestamp = conditional_join.withColumn(
+        "_rank", row_number().over(window)
+    ).filter(col("_rank") == 1)
+
+    return filter_most_recent_feature_timestamp.select(
+        entity.columns + [f"{feature_prefix}{feature}" for feature in features]
+    )
+
+
+def join_entity_to_feature_tables(
+    query_conf: List[Dict[str, Any]], entity: DataFrame, tables: Dict[str, DataFrame]
+) -> DataFrame:
+    """Perform as of join between entity and multiple feature table. Returns a DataFrame.
+
+    Args:
+        query_conf (List[Dict[str, Any]]):
+            Query configuration.
+        entity (DataFrame):
+            Entity dataframe. Must contain the column event_timestamp.
+        tables (Dict[str, DataFrame]):
+            Map of feature table name to Spark DataFrame.
+
+    Returns:
+        DataFrame: Join result.
+
+    """
+    joined = entity
+    for query in query_conf:
+        joined = as_of_join(
+            joined,
+            query["join"],
+            tables[query["table"]],
+            query["features"],
+            feature_prefix=f"{query['table']}__",
+            max_age=query.get("max_age"),
+        )
+    return joined
+
+
+def batch_retrieval(spark: SparkSession, conf: Dict) -> DataFrame:
+    """Retrieve batch features based on given configuration.
+
+    Args:
+        spark (SparkSession):
+            Spark session.
+        conf (Dict):
+            Configuration for the retrieval job, in json format. Sample configuration as follows:
+
+            sample_conf = {
```

Comment:
This can be moved into an example https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html